crates/core/tests/connection_cap.rs

```diff
@@ -0,0 +1,36 @@
+//! Minimal repro harness for connection-cap enforcement.
+//!
+//! This test spins up a tiny network (2 gateways + 6 peers) with a low max-connections
+//! setting (min=5, max=6) and waits for connectivity. It then inspects diagnostics to
+//! ensure no peer reports more than `max` connections. This is intended to quickly catch
+//! admission/cap bypass regressions without running the full soak.
+
+use freenet_test_network::{BuildProfile, FreenetBinary, NetworkBuilder};
```

Comment:
Is very frustrating that we already talked about not having to bring out of tree test harness , but we still have done it.

Reply:
Understood. The soak/connection-cap tests currently rely on the `freenet-test-network` harness for churn and diagnostics; I kept it to get the coverage landed with this transient stack. I can follow up to inline the needed pieces (or move the harness into this repo) so we’re not depending on out-of-tree test tooling long-term.